File: python/src/p_1342_nums_of_steps.py
class Solution:
    def numberOfSteps(self, num: int) -> int:
        if num == 0:
            return 0
        return num.bit_length() - 1 + num.bit_count()


if __name__ == '__main__':
    s = Solution()
    n = 14
    print(f'n={n}, steps={s.numberOfSteps(n)}')
    assert s.numberOfSteps(n) == 6
    print('-' * 10)

    n = 8
    print(f'n={n}, steps={s.numberOfSteps(n)}')
    assert s.numberOfSteps(n) == 4
    print('-' * 10)

    n = 123
    print(f'n={n}, steps={s.numberOfSteps(n)}')
    assert s.numberOfSteps(n) == 12
    print('-' * 10)

    n = 0
    print(f'n={n}, steps={s.numberOfSteps(n)}')
    assert s.numberOfSteps(n) == 0
    print('-' * 10)

    n = 1
    print(f'n={n}, steps={s.numberOfSteps(n)}')
    assert s.numberOfSteps(n) == 1
    print('-' * 10)

File: typescript/src/p_1342_nums_of_steps.ts
function numberOfSteps(num: number): number {
  if (num === 0) return 0;

  const bitLength = Math.floor(Math.log2(num)) + 1;

  let bitCount = 0;
  let n = num;
  while (n > 0) {
    bitCount += n & 1;
    n >>>= 1;
  }

  return (bitLength - 1) + bitCount;
};

function tests() {
    let num = 14;
    let expected = 6;
    let result = numberOfSteps(num);
    console.assert(result === expected, `Test case 1 failed: numberOfSteps(${num}) = ${result}, expected ${expected}`);
    console.log(`Test case 1 passed: numberOfSteps(${num}) = ${result}`);

    num = 8;
    expected = 4;
    result = numberOfSteps(num);
    console.assert(result === expected, `Test case 2 failed: numberOfSteps(${num}) = ${result}, expected ${expected}`);
    console.log(`Test case 2 passed: numberOfSteps(${num}) = ${result}`);

    num = 123;
    expected = 12;
    result = numberOfSteps(num);
    console.assert(result === expected, `Test case 3 failed: numberOfSteps(${num}) = ${result}, expected ${expected}`);
    console.log(`Test case 3 passed: numberOfSteps(${num}) = ${result}`);

    num = 0;
    expected = 0;
    result = numberOfSteps(num);
    console.assert(result === expected, `Test case 4 failed: numberOfSteps(${num}) = ${result}, expected ${expected}`);
    console.log(`Test case 4 passed: numberOfSteps(${num}) = ${result}`);
}

tests();
